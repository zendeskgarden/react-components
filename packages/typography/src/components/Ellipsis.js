/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { StyledEllipsis } from '../styled';

/**
 * A component that automatically includes a native `title` attribute and any
 * text-overflow styling. All other props are spread onto the element.
 */
const Ellipsis = React.forwardRef(({ children, title, tag, ...other }, ref) => {
  let textContent = null;

  if (title !== undefined) {
    textContent = title;
  } else if (typeof children === 'string') {
    textContent = children;
  }

  return (
    <StyledEllipsis as={tag} ref={ref} title={textContent} {...other}>
      {children}
    </StyledEllipsis>
  );
});

Ellipsis.propTypes = {
  /**
   * Optional override for the auto-generated `title` attribute
   */
  title: PropTypes.string,
  /** Any valid element for the styled component */
  tag: PropTypes.any,
  children: PropTypes.string
};

Ellipsis.defaultProps = {
  tag: 'div'
};

export default Ellipsis;
