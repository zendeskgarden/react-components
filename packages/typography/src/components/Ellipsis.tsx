/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { HTMLAttributes } from 'react';
import { StyledEllipsis } from '../styled';

interface IEllipsisProps extends HTMLAttributes<HTMLSpanElement> {
  /**
   * Optional override for the auto-generated `title` attribute
   */
  title?: string;
  /** Any valid element for the styled component */
  tag?: any;
}

/**
 * A component that automatically includes a native `title` attribute and any
 * text-overflow styling. All other props are spread onto the element.
 */
const Ellipsis = React.forwardRef<HTMLDivElement, IEllipsisProps>(
  ({ children, title, tag, ...other }, ref) => {
    let textContent = undefined;

    if (title !== undefined) {
      textContent = title;
    } else if (typeof children === 'string') {
      textContent = children;
    }

    return (
      <StyledEllipsis as={tag} ref={ref} title={textContent} {...other}>
        {children}
      </StyledEllipsis>
    );
  }
);

Ellipsis.defaultProps = {
  tag: 'div'
};

export default Ellipsis;
