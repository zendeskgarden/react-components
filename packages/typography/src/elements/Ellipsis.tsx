/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React, { HTMLAttributes } from 'react';
import PropTypes from 'prop-types';
import { StyledEllipsis } from '../styled';

interface IEllipsisProps extends HTMLAttributes<HTMLSpanElement> {
  /**
   * Optional override for the auto-generated `title` attribute
   */
  title?: string;
  /** Any valid element for the styled component */
  tag?: any;
}

/**
 * A component that automatically includes a native `title` attribute and any
 * text-overflow styling. Accepts all standard attributes and events for the
 * provided `tag`.
 */
const Ellipsis: React.FunctionComponent<
  IEllipsisProps & React.RefAttributes<HTMLDivElement>
> = React.forwardRef<HTMLDivElement, IEllipsisProps>(({ children, title, tag, ...other }, ref) => {
  let textContent = undefined;

  if (title !== undefined) {
    textContent = title;
  } else if (typeof children === 'string') {
    textContent = children;
  }

  return (
    <StyledEllipsis as={tag} ref={ref} title={textContent} {...other}>
      {children}
    </StyledEllipsis>
  );
});

Ellipsis.propTypes = {
  title: PropTypes.string,
  tag: PropTypes.any
};

Ellipsis.defaultProps = {
  tag: 'div'
};

export default Ellipsis;
