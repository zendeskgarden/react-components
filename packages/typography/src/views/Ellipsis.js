/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { retrieveTheme, isRtl } from '@zendeskgarden/react-theming';

const COMPONENT_ID = 'typography.ellipsis';

const StyledEllipsis = styled.div.attrs({
  'data-garden-id': COMPONENT_ID,
  'data-garden-version': PACKAGE_VERSION
})`
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;

  direction: ${props => (isRtl(props) ? 'rtl' : 'ltr')};

  ${props => retrieveTheme(COMPONENT_ID, props)};
`;

/**
 * A component that automatically includes a native `title` attribute
 * and any text-overflow styling.
 *
 * All other props are spread onto the element.
 */
function Ellipsis({ children, title, tag, ...other }) {
  const CustomTagEllipsis = StyledEllipsis.withComponent(tag);

  let textContent = null;

  if (title !== undefined) {
    textContent = title;
  } else if (typeof children === 'string') {
    textContent = children;
  }

  return (
    <CustomTagEllipsis title={textContent} {...other}>
      {children}
    </CustomTagEllipsis>
  );
}

Ellipsis.propTypes = {
  /**
   * Optional override for the auto-generated `title` attribute
   */
  title: PropTypes.string,
  /** Any valid element for the styled component */
  tag: PropTypes.any,
  children: PropTypes.string
};

Ellipsis.defaultProps = {
  tag: 'div'
};

export default Ellipsis;
