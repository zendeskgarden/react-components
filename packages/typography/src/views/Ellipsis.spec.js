/**
 * Copyright Zendesk, Inc.
 *
 * Use of this source code is governed under the Apache License, Version 2.0
 * found at http://www.apache.org/licenses/LICENSE-2.0.
 */

import React from 'react';
import { mountWithTheme } from '@zendeskgarden/react-testing';
import Ellipsis from './Ellipsis';

const Example = props => <Ellipsis {...props}>Hello world</Ellipsis>;

describe('Ellipsis', () => {
  it('applies correct styling by default', () => {
    const wrapper = mountWithTheme(<Example />);

    expect(wrapper).toMatchSnapshot();
  });

  it('overrides title attribute if provided', () => {
    const wrapper = mountWithTheme(<Example title="Custom title" />);

    expect(wrapper).toMatchSnapshot();
  });

  it('applies correct styling with RTL locale', () => {
    const wrapper = mountWithTheme(<Example />, { rtl: true });

    expect(wrapper).toMatchSnapshot();
  });
});
