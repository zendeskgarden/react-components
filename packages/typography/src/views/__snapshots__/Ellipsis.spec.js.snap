// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Ellipsis applies correct styling by default 1`] = `
.c0 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  direction: ltr;
}

<Example>
  <Ellipsis
    tag="div"
  >
    <Ellipsis__StyledEllipsis
      title="Hello world"
    >
      <div
        className="-div c0"
        data-garden-id="typography.ellipsis"
        data-garden-version="version"
        title="Hello world"
      >
        Hello world
      </div>
    </Ellipsis__StyledEllipsis>
  </Ellipsis>
</Example>
`;

exports[`Ellipsis applies correct styling with RTL locale 1`] = `
.c0 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  direction: rtl;
}

<Example>
  <Ellipsis
    tag="div"
  >
    <Ellipsis__StyledEllipsis
      title="Hello world"
    >
      <div
        className="-div c0"
        data-garden-id="typography.ellipsis"
        data-garden-version="version"
        title="Hello world"
      >
        Hello world
      </div>
    </Ellipsis__StyledEllipsis>
  </Ellipsis>
</Example>
`;

exports[`Ellipsis overrides title attribute if provided 1`] = `
.c0 {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  direction: ltr;
}

<Example
  title="Custom title"
>
  <Ellipsis
    tag="div"
    title="Custom title"
  >
    <Ellipsis__StyledEllipsis
      title="Custom title"
    >
      <div
        className="-div c0"
        data-garden-id="typography.ellipsis"
        data-garden-version="version"
        title="Custom title"
      >
        Hello world
      </div>
    </Ellipsis__StyledEllipsis>
  </Ellipsis>
</Example>
`;
